aip/0151.md

```diff
@@ -0,0 +1,86 @@
+---
+aip:
+  id: 151
+  state: reviewing
+  created: 2019-07-25
+permalink: /151
+redirect_from:
+  - /0151
+---
+
+# Long-running operations
+
+Occasionally, an API may need to expose a method that takes a significant
+amount of time to complete. In these situations, it is often a poor user
+experience to simply block while the task runs; rather, it is better to return
+some kind of promise to the user and allow the user to check back in later.
+
+The long-running operations pattern is roughly analogous to a [Python
+`Future`][], or a [Node.js `Promise`][]. Essentially, the user is given a token
+that can be used to track progress and retrieve the result.
+
+## Guidance
+
+Individual API methods that might take a significant amount of time to complete
+**should** return a [`google.longrunning.Operation`][lro] object instead of the
+ultimate response message.
+
+```proto
+// Write a book.
+rpc WriteBook(WriteBookRequest) returns (google.longrunning.Operation) {
+  option (google.api.http) = {
+    post: "/v1/{parent=shelves/*}/books}:write"
+    body: "*"
+  };
+  option (google.longrunning.operation_info) = {
+    response_type: "WriteBookResponse"
+    metadata_type: "WriteBookMetadata"
+  };
+}
+```
+
+- The response type **must** be `google.longrunning.Operation`. The `Operation`
+  proto definition **must not** be copied into individual APIs.
+- The method **must** include a `google.longrunning.operation_info` annotation
+  defining the response and metadata types.
+  - The response and metadata types **must** be defined in the file where the
+    RPC appears, or a file imported by that file.
+  - The metadata type **should not** be `google.protobuf.Empty`, as this
```

Comment:
Maybe say that the response type must not be empty either?

Reply:
Done.